pkg/quota/image/usage.go

```diff
@@ -10,139 +13,400 @@ import (
+			ns = objRef.Namespace
+		}
+
+		isName := nameParts[0]
+		is, err := c.getImageStream(ns, isName)
+		if err != nil {
+			return imageapi.DockerImageReference{}, fmt.Errorf("failed to get imageStream for ImageStreamTag %s/%s: %v", ns, objRef.Name, err)
+		}
+
+		event := imageapi.LatestTaggedImage(is, nameParts[1])
+		if event == nil || len(event.DockerImageReference) == 0 {
+			return imageapi.DockerImageReference{}, fmt.Errorf("%q is not currently pointing to an image, cannot use it as the source of a tag", objRef.Name)
+		}
+		return imageapi.ParseDockerImageReference(event.DockerImageReference)
+
+	case "DockerImage":
+		managedByOS, ref := imageReferenceBelongsToInternalRegistry(objRef.Name)
+		if !managedByOS {
+			return imageapi.DockerImageReference{}, fmt.Errorf("DockerImage %s does not belong to internal registry", objRef.Name)
 		}
+		return ref, nil
 	}
 
-	return
+	return imageapi.DockerImageReference{}, fmt.Errorf("unsupported object reference kind %s", objRef.Kind)
+}
+
+// getImageStream gets an image stream object from etcd and caches the result for the following queries.
+func (c *GenericImageStreamUsageComputer) getImageStream(namespace, name string) (*imageapi.ImageStream, error) {
+	key := fmt.Sprintf("%s/%s", namespace, name)
+	if is, exists := c.imageStreamCache[key]; exists {
+		return is, nil
+	}
+	is, err := c.osClient.ImageStreams(namespace).Get(name)
+	if err == nil {
+		c.imageStreamCache[key] = is
+	}
+	return is, err
+}
+
+// getImage gets image object from etcd and caches the result for the following queries.
+func (c *GenericImageStreamUsageComputer) getImage(name string) (*imageapi.Image, error) {
+	if image, exists := c.imageCache[name]; exists {
+		return image, nil
+	}
+	image, err := c.osClient.Images().Get(name)
+	if err == nil {
+		c.imageCache[name] = image
+	}
+	return image, err
+}
+
+// getImageStreamImage gets an image belonging to a given image stream object from etcd and caches the result for the following queries.
+func (c *GenericImageStreamUsageComputer) getImageStreamImage(namespace, name string) (*imageapi.Image, error) {
+	nameParts := strings.SplitN(name, "@", 2)
+	if len(nameParts) != 2 {
+		return nil, fmt.Errorf("failed to parse image stream image name %q, expected exactly one '@'", name)
+	}
+	image, exists := c.imageCache[nameParts[1]]
+	if exists {
+		return image, nil
+	}
+	isi, err := c.osClient.ImageStreamImages(namespace).Get(nameParts[0], nameParts[1])
+	if err == nil {
+		image = &isi.Image
+		c.imageCache[name] = image
+	}
+	return image, err
+}
+
+// listImageStreams returns a list of image streams of the given namespace and caches them for later access.
+func (c *GenericImageStreamUsageComputer) listImageStreams(namespace string) (*imageapi.ImageStreamList, error) {
+	iss, err := c.osClient.ImageStreams(namespace).List(kapi.ListOptions{})
+	if err == nil {
+		for _, is := range iss.Items {
+			c.imageStreamCache[fmt.Sprintf("%s/%s", namespace, is.Name)] = &is
```

Comment:
This is an even bigger problem - you're effectively maintaining a completely custom cache of image streams and images... but this is going to be run in a context where we will already have caches of image streams and images.  So this effectively doubles the memory usage of the controller, which is a significant burden.

This needs to be refactored to share a real client.Cache with at least one other controller in this loop.